### PR TITLE
feat(wct-runner): Add plugin for the WCT test runner

### PR DESCRIPTION
**Pull request by nicojs**
Add the `stryker-wct-runner` package. A plugin to use [web-component-tester](https://github.com/Polymer/tools/tree/master/packages/web-component-tester) with Stryker. It allows WCT to use the default `wct.conf.json` configuration, or you can configure it with the `wct` property in your stryker.conf.js file. For example: 

```js
// Full example
wct: {
    configFile: 'alternative/wct/configuration.json',
    verbose: true,
    npm: true,
    plugins: {
        local: {
            browsers: ["chrome"],
            skipSeleniumInstall: true,
            browserOptions: {
                chrome: ["window-size=1920,1080", "headless", "disable-gpu"],
                firefox: ["--headless"]
            }
        }
    }
}
```

For more information, see the readme.md.

Some more points:

* `web-component-tester` is a peer dependency. Users are responsible to install a correct version. 
* WCT can be chatty. By default, all log messages are ignored. Setting `wct.verbose` to `true` will forward all wct logging to Stryker's logging system.
* WCT uses Node's `promisify` implementation directly. As such, it will only work from node 8 onward. For that reason, wct integration tests don't run on our node 6 build. 
* The option `persistent: false` is forced as wct option, because `persistent: true` keeps the browser open and Stryker cannot read the test results. See https://github.com/Polymer/tools/issues/796 for more details.

- [x] Add readme
- [x] Add some more wct related logging
- [x] Make sure wct can run on travis
- [x] Throw error if coverage analysis != 'off'
- [x] Create integration tests
- [x] Test on an actual polymer project
- [x] Add unit tests where needed
- [x] Disable "persistent" boolean

Issues to report:
- [x] Ship WCT with header files, instead of actual ts files.
- [x] Allow WCT to reuse the same browser for a massive performance boost
- [x] Set the "typings" field in the WCT package.json
- [x] Add `Context` class to be exported via the public interface

We might also want to do something about this:

* Allow WCT to exit cleanly. Right now keeps nodejs process open: https://pastebin.com/gTGDMaPx

### CODE REVIEW

File: packages/stryker-wct-runner/src/WctReporter.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+import { TestResult, TestStatus } from 'stryker-api/test_runner';
+
+const TEST_START_EVENT = 'test-start';
+const TEST_END_EVENT = 'test-end';
+
+export default class WctReporter {
+  public results: TestResult[] = [];
+  private before = new Date();
+
+  constructor(private readonly emitter: EventEmitter) {
```
[the remark this answers is not in the document]
Comment (simondel) [reply packages/stryker-wct-runner/src/WctReporter.ts]: I just noticed that you also call this `context` in the spec files for WctReporter

File: packages/stryker-wct-runner/src/WctTestRunner.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  private readonly context: Context;
+  private readonly logger: WctLogger;
+  private readonly log = getLogger(WctTestRunner.name);
+
+  constructor(runnerOptions: { strykerOptions: StrykerOptions }) {
+    if (runnerOptions.strykerOptions.coverageAnalysis !== 'off') {
+      throw new Error(`Coverage analysis "${runnerOptions.strykerOptions.coverageAnalysis}" is not (yet) supported by the WCT test runner plugin. Please set \`coverageAnalysis: "off"\` in your stryker.conf.js file.`);
+    }
+    this.log.debug('Running wct version %s from %s', require(`${WCT_PACKAGE}/package.json`).version, require.resolve(WCT_PACKAGE));
+    this.context = new Context(Object.assign({}, runnerOptions.strykerOptions.wct, FORCED_WCT_OPTIONS));
```
[the remark this answers is not in the document]
Comment (nicojs) [reply packages/stryker-wct-runner/src/WctTestRunner.ts]: Agreed, moving context stuff to a method.

File: packages/stryker-wct-runner/src/WctTestRunner.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
+import { TestRunner, RunResult, RunStatus } from 'stryker-api/test_runner';
+import { Context } from 'web-component-tester/runner/context';
+import { steps } from 'web-component-tester';
+import { StrykerOptions } from 'stryker-api/core';
+import WctReporter from './WctReporter';
+import WctLogger from './WctLogger';
+import { getLogger } from 'stryker-api/logging';
+const WCT_PACKAGE = 'web-component-tester';
+const FORCED_WCT_OPTIONS = Object.freeze({
```
[the remark this answers is not in the document]
Comment (nicojs) [reply packages/stryker-wct-runner/src/WctTestRunner.ts]: I think it's clean to keep them at the top, not polluting the constructor.

File: packages/stryker-wct-runner/src/WctReporter.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    emitter.on(
```
[the remark this answers is not in the document]
Comment (nicojs) [reply packages/stryker-wct-runner/src/WctReporter.ts]: Agreed, fixed

File: packages/stryker-wct-runner/src/WctReporter.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        } else {
+          return undefined;
+        }
+      default:
+        return [error.toString()];
+    }
+  }
+
+  private testNamePartsToString(testNameParts: string[]): string {
+    return testNameParts.reduce((prev, current, index) => {
```
[the remark this answers is not in the document]
Comment (nicojs) [reply packages/stryker-wct-runner/src/WctReporter.ts]: Because I'm over complicating things. Fixed

[the remark this answers is not in the document]
Comment (nicojs) [reply packages/stryker-wct-runner/src/WctReporter.ts]: EDIT: this is actually important:  `// First part is the file name`. So I'm removing the first part.

[the remark this answers is not in the document]
Comment (nicojs) [reply packages/stryker-wct-runner/src/WctReporter.ts]: Saved by a unit test

File: packages/stryker-wct-runner/src/WctTestRunner.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (simondel) [reply packages/stryker-wct-runner/src/WctTestRunner.ts]: True, but you could also place them inside of the class

File: packages/stryker-wct-runner/src/WctReporter.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (simondel) [reply packages/stryker-wct-runner/src/WctReporter.ts]: Then why not use `return testNameParts.splice(1).join(' ').trim();`?

[the remark this answers is not in the document]
Comment (nicojs) [reply packages/stryker-wct-runner/src/WctReporter.ts]: performance?

[the remark this answers is not in the document]
Comment (nicojs) [reply packages/stryker-wct-runner/src/WctReporter.ts]: 🙈 will fix

File: packages/stryker-wct-runner/src/WctReporter.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (nicojs) [reply packages/stryker-wct-runner/src/WctReporter.ts]: Hmm ok

File: packages/stryker-wct-runner/src/WctTestRunner.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (nicojs) [reply packages/stryker-wct-runner/src/WctTestRunner.ts]: Where? I don't want them to class properties, right?

**Pull request merged**